doc/internal/consensus/Consensus-Protocol.md

```diff
@@ -0,0 +1,80 @@
+* Election is repeated after an election timeout until:
+  1. The server wins the election
+  2. A peer establishes itself as leader.
+  3. Election timer times out or a split vote occurs (leading to no leader) and the process will be repeated.
+ * Election win: Majority votes in the term. (More details in safety) The state of the winner is now `Leader` and the others are `Followers`.
+ * Maintaining leaders reign: The leader sends `heartbeats` to all servers to establish its reign. This also checks whether other servers are alive based on the response and informs other servers that the leader still is alive too. If the servers do not get timely heartbeat messages, they transform from the `follower` state to `candidate` state.
+ * Transition from working state to Election happens when a leader fails.
+ * Maintaining sanity: While waiting for votes, if a `AppendEntriesRPC` is received by the server, and the term of the leader is greater than of equal to the "waiter"'s term, the leader is considered to be legitimate and the waiter becomes a follower of the leader. If the term of the leader is lesser, it is rejected.
+ * The split vote problem: Though not that common, split votes can occur. To make sure this doesnt continue indefinitely, election timeouts are randomised, making the split votes less probable.
+ 
+
+#### Implementation
+
+* A separate `interal/raft` folder will have a raft implementation which provides APIs for each server to call.
+* 
+### Log Replication
+
+#### Spec
+
+* Pre-log replication: Once a leader is elected, it starts servicing the client. The leader appends a new request to its `New Entry` log then issues `AppendEntriesRPC` in parallel to all its peers. 
+* Successful log: When all logs have been applied successfully to all follower machines, the leader applies the entry to its state machine and returns the result to the client.
+* Repeating `AppendEntries`: `AppendEntriesRPC` are repeated indefinitely until all followers eventually store all log entries.
+* Log entry storage: Log entries are a queue of state machine commands which are applied to that particular state machine. Log entries are associated with a term number to indicate the term of application of that log along with an integer index to identify a particular logs position.
+* Committed entry: A log entry is called committed once its replicated on the majority of the servers in the cluster. Once an entry is committed, it commits all the previous entries in the leaders log, including the entries created by the previous leaders.
+* The  leader keeps track of the highest known index that it knows is committed and it is included in all the future `AppendEntriesRPC` (including heartbeats) to inform other servers.
+* Theres some issue about log committing - paragraph says its committed when its applied everywhere and also says its applied everywhere once its committed. 
```

Comment:
what issue in log commiting, I think in doc we should put how we are going to design and what we are going to do for given problem rather talk about problem. We can discuss find it's solution or working solution

Reply:
Outdated.